keps/sig-cli/NNNN-kustomize-components.md

```diff
@@ -0,0 +1,676 @@
+├── components
+│   ├── external_db
+│   │   ├── configmap.yaml
+│   │   ├── dbpass.txt
+│   │   ├── deployment.yaml
+│   │   └── kustomization.yaml
+│   ├── ldap
+│   │   ├── configmap.yaml
+│   │   ├── deployment.yaml
+│   │   ├── kustomization.yaml
+│   │   └── ldappass.txt
+│   └── recaptcha
+│       ├── deployment.yaml
+│       ├── kustomization.yaml
+│       ├── secret_key.txt
+│       └── site_key.txt
+└── overlays
+    ├── community
+    │   └── kustomization.yaml
+    ├── dev
+    │   └── kustomization.yaml
+    └── enterprise
+        └── kustomization.yaml
+```
+
+With this structure, we can create the YAML files for each deployment as
+follows:
+
+```shell
+kustomize build overlays/community
+kustomize build overlays/enterprise
+kustomize build overlays/dev
+```
+
+### Notes/Constraints/Caveats (optional)
+
+The distinction between components and overlays is evident when transformers
+are involved, but not when a component adds resources only. This may confuse
+users who will not know what to use in this case, and must be communicated
+clearly in the docs.
+
+### Risks and Mitigations
+
+Since a component is a special type of a kustomization, that ensures
+serialization of resources, the same logic and restrictions apply here as well.
+This means that the threat model of components is no different from the threat
+model for overlays or other kustomizations.
+
+## Design Details
+
+When Kustomize processes a `Kustomization` file it uses a `ResourceAccumulator`
+(RA) object, which represents the accumulated state of the processing up to
+that point (including files, patches, transformers, variables, etc).  The RA is
+initially empty, and then Kustomize:
+
+1. Adds each of its resources (in left-to-right order) to the RA
+   (`kusttarget.accumulateResources`):
+  * Adds files to the RA directly (`accumulateFile`).
+  * Recursively processes Kustomizations, each starting with a new empty RA,
+    and with the result merged in to the RA of the parent
+    (`accumulateDirectory` and `resaccumulator.MergeAccumulator`).
+2. Adds/merges its CRDs, generators, transformers, and variables to the RA.
+  * These modifications can only be applied to entities within the RA - it is
+    an error otherwise.  This means that sibling Kustomizations are entirely
+    independent of each other.
+  * Kustomizations with the same base will cause an error when the resources
+    are accumulated, unless their GVKN is altered.
+
+`Components` are processed between step 1 and 2, and in a similar manner to
+Kustomizations, but instead of starting with an empty RA, they take the RA from
+their parent.  The component is therefore able to modify everything within the
+RA, which will contain all of the parent's resources and the result of
+processing all of the earlier components in the `components:` list. Nested
+components will work in the same way - ownership of the parent's RA will simply
+be passed down.
+
+This was implemented by changing the signature of
```

Comment:
I think that in these types of documents, we tend to use the present tense, similar to Git commits. For instance this is more common:

This was implemented by changing -> We implement this by changing

Reply:
Looks good to me!

Reply:
I went a bit overboard and reworded some other stuff as well :nerd_face: . Tell me if you're ok with the changes (you can see the diff [here](https://github.com/arrikto/kubernetes-enhancements/pull/1/commits/4de490b4e1021d4624f9609cf6521c69c51f91ea)).

Reply:
Changes look good!